### ISSUE TITLE
Deadlock and AMQP reconnections when prefetched messages exceed max workers

### ISSUE THREAD

**Issue by mattbennett** [nameko/nameko#428]
The interaction of the AMQP `prefetch_count` and Nameko's worker pool can result in strange behaviour in some circumstances.

When the number of unack'd messages exceeds the worker pool size, the QueueConsumer thread blocks waiting for a worker. Consequently, any outstanding message acks are not processed. This can result in deadlock if the running workers use RPC proxies.

The QueueConsumer thread is also responsible for sending heartbeats. If it thread ends up waiting for the worker pool longer two heartbearts, the broker will the connection. Any in-flight messages will be re-queued and the workers duplicated. Any currently running workers that are waiting for an RPC reply will raise "disconnected waiting for reply" errors upon reconnection.

This bug has been present for a long time but is easier to trigger as a result of the change in https://github.com/nameko/nameko/pull/419.

There is a simple fix, which I will raise in a PR now.

This raises the question of whether the worker pool is a useful construct or not. I am still forming an opinion but my current thoughts are:

* 10 workers is way too low for the default
* perhaps we should remove the restriction altogether, and just let throughput be constrained by available resources
* it may be useful for some _entrypoints_ to be constrained (i.e. only run x of these types of workers concurrently) but having a global constraint shared amongst all types of workers causes confusion without really solving a problem

**Comment by sideboltconor** [comment]
Hey Matt,

You must be reading my mind as I have been struggling with these deadlocks today. Last night one of our servers died which created more load on our other nodes, which then starting raising deadlock errors like crazy. 

Our solution to the deadlocks has always been to just deploy more nodes. I've never fully understood the right max_worker count for our deployment. Our gateway (web socket service) is set to 1,000 max_workers and all other services are set to 100. These were just guesses, and the reason we set the web socket service to 1000 is that it seemed the most susceptible to deadlocks. 

Anyway, a fix to this issues would be greatly appreciated by our team!

Thanks,
Conor

**Comment by mattbennett** [comment]
Hi Conor,

Sorry to hear about this! We've also struggled with hung services occasionally at Student.com, but the culprit always seemed to be RPC proxies waiting for replies (usually due to a downstream service that was overloaded). That blinded us to this underlying problem, which has existed in some form for a long time but is made much worse by the 2.5.3 release. I've tried to capture the various failure cases in the tests in the PR.

Which version of Nameko are your services running?

As I said above, I'm coming to the conclusion that `max_workers` is an unhelpful feature, at least for entrypoints where you can otherwise constrict the amount of concurrency (e.g. the AMQP entrypoints, although the implementation there is far from perfect too). It's probably sensible to set it to a very high number for now, unless you have a particular reason to throttle a service.

The fix is coming soon, and it's embarrassingly simple 😳

**Comment by sideboltconor** [comment]
Hi Matt,

Simple fixes are the best fixes 👍 

We are on 2.5.3, we updated as soon as it hit hoping the max_workers +1 would solve our issues.

So if we set the max_workers to something "very high", does that mean that the Nameko service will prefetch that many tasks from rabbit? Meaning if I have 1000 tasks sitting in my queue and my max_workers is set to 1000 will the first nameko host grab all 1000?

Cheers,
Conor

**Comment by mattbennett** [comment]
Being on 2.5.3 explains why you're suddenly experiencing them more often.

Sorry, my advice to jack up the worker count was a little cavalier. You'll separately need to set the `prefetch_count` to something sensible, otherwise (as you pointed out) the AMQP entrypoints will try to consume that many messages in one go.

It is a bit of a faff but you can override the prefetch count by subclassing the `QueueConsumer`, and the chain of Entrypoints / Extensions that use it:

``` python
# limited.py
from nameko.messaging import QueueConsumer as NamekoQueueConsumer
from nameko.rpc import Rpc as NamekoRpc
from nameko.rpc import RpcConsumer as NamekoRpcConsumer


class QueueConsumer(NamekoQueueConsumer):
    prefetch_count = 10


class RpcConsumer(NamekoRpcConsumer):
    queue_consumer = QueueConsumer()


class Rpc(NamekoRpc):
    rpc_consumer = RpcConsumer()


rpc = Rpc.decorator


class Service:
    name = "your-service"

    @rpc
    def method(self):
        pass
```

You'd need to do the same for any `consume` and `event_handler` entrypoints too.

Run this with the following config

``` yaml
# config.yaml
max_workers: 1000
AMQP_URI: amqp://guest:guest@localhost:5672/
```

And you'll see the RPC consumer is limited to 10 but has a pool size of 1000:

```
$ nameko run limited --config config.yaml --backdoor 3000
starting services: your-service
backdoor server listening on 127.0.0.1:3000
Connected to amqp://guest:**@127.0.0.1:5672//
```

```
$ nameko backdoor 3000
Python 2.7.10 (default, Jul  6 2015, 15:19:48)
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> ct = list(runner.containers)[0]
>>> ct._worker_pool.size
1000
```

**Comment by sideboltconor** [comment]
Thank you Matt, this is a great help.

**Comment by mattbennett** [comment]
Fixed in 2.5.4
